ledger/forest/forest_test.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package forest
+
+import (
+	"bytes"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/encoding"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// TestLeftEmptyInsert tests inserting a new value into an empty sub-trie:
+//   1. we first construct a baseTrie holding a couple of values on the right branch [~]
+//   2. we update a previously non-existent register on the left branch (X)
+// We verify that values for _all_ paths in the updated Trie have correct payloads
+func TestLeftEmptyInsert(t *testing.T) {
+	//////////////////////
+	//     insert X     //
+	//       ()         //
+	//      /  \        //
+	//    (X)  [~]      //
+	//////////////////////
+
+	forest := New()
+
+	// path: 1000...
+	p1 := pathByUint8s([]uint8{uint8(129), uint8(1)})
+	v1 := payloadBySlices([]byte{'A'}, []byte{'A'})
+
+	// path: 1100...
+	p2 := pathByUint8s([]uint8{uint8(193), uint8(1)})
+	v2 := payloadBySlices([]byte{'B'}, []byte{'B'})
+
+	paths := []ledger.Path{p1, p2}
+	payloads := []*ledger.Payload{v1, v2}
+	update := &ledger.TrieUpdate{RootHash: trie.EmptyTrieRootHash(), Paths: paths, Payloads: payloads}
+	baseRoot, err := forest.Update(update)
+	require.NoError(t, err)
+
+	fmt.Println("trees:", len(forest.steps))
+	for hash, tr := range forest.steps {
+		fmt.Printf("%x: %d\n", hash, len(tr.tree.Leaves()))
+	}
```

Comment:
Do we want writing to `STDOUT` in tests?

Reply:
Nope but as stated before, this file is not intended to be reviewed for now, I only pushed because there was a missing important fix 😅